libraries/AP_IOMCU/AP_IOMCU.cpp

```diff
@@ -613,4 +631,42 @@ void AP_IOMCU::update_safety_options(void)
     }
 }
 
+/*
+  check ROMFS firmware against CRC on IOMCU, and if incorrect then upload new firmware
+ */
+bool AP_IOMCU::check_crc(void)
+{
+    // flash size minus 4k bootloader
+	const uint32_t flash_size = 0x10000 - 0x1000;
+    uint32_t fw_size;
+    
+    fw = AP_ROMFS::find_file(fw_name, fw_size);
+    if (!fw) {
+        hal.console->printf("failed to find %s\n", fw_name);
+        return false;
+    }
+    uint32_t crc = crc_crc32(0, fw, fw_size);
+
+    // pad to max size
+	while (fw_size < flash_size) {
+		uint8_t b = 0xff;
+		crc = crc_crc32(crc, &b, 1);
+		fw_size++;
+	}
+
+    uint32_t io_crc = 0;
+    if (read_registers(PAGE_SETUP, PAGE_REG_SETUP_CRC, 2, (uint16_t *)&io_crc) &&
+        io_crc == crc) {
+        hal.console->printf("IOMCU: CRC ok\n");
+        return true;
+    }
+
+    const uint16_t magic = REBOOT_BL_MAGIC;
```

Comment:
Isn't rebooting to the bootloader part of `upload_fw`'s domain rather then `check_crc`?